### PR TITLE
Added Isrlu activation function

### PR DESCRIPTION
**Pull request by PranavReddyP16**
added isrlu activation funtion

### CODE REVIEW

File: src/mlpack/methods/ann/activation_functions/isrlu_function.cpp [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+ *      &=& x                     if x>=0
+ *
+ * f'(x) &=& cube(1/sqrt(1 + alpha*x*x)) if x<0
+ *       &=& 1                           if x>=0
+ * @f
+ */
+
+class ISRLUFunction
+{
+
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/activation_functions/isrlu_function.cpp]: No need of an empty line here.

File: src/mlpack/methods/ann/activation_functions/isrlu_function.cpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+  /**
+   * Computes the ISRLU function
+   * 
+   * @param x Input data.
+   * @param alpha paramter, default value = 0.1
+   * @return f(x)
+   */
+  static double Fn(const double x, const double alpha = 0.1)
+  {
+    if (x<0)
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/activation_functions/isrlu_function.cpp]: space between operators.

File: src/mlpack/methods/ann/activation_functions/isrlu_function.cpp [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  /**
+   * Computes the ISRLU function using a matrix as a input
+   * 
+   * @param x Input data.
+   * @param y The resulting output activation
+   * @param alpha parameter, default value = 0.1
+   */
+  template<typename eT>
+  static void Fn(const arma::Mat<eT>& x,
+             arma::Mat<eT>& y,
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/activation_functions/isrlu_function.cpp]: Could you fix the spacing here. Refer any other activation function.

File: src/mlpack/methods/ann/activation_functions/isrlu_function.cpp [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+             const double alpha = 0.1)
+  {
+    if (x<0)
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/activation_functions/isrlu_function.cpp]: Same as above.

File: src/mlpack/methods/ann/activation_functions/isrlu_function.cpp [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+   * @param alpha parameter, default value = 0.1
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Fn(const InputVecType& x,
+                 OutputVecType& y,
+                 const double alpha = 0.1)
+  {
+    y.set_size(arma::size(x));
+
+    for (size_t i = 0; i < x.n_elem; i++)
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/activation_functions/isrlu_function.cpp]: Use the same code as above. PS you are already using a template.

File: src/mlpack/methods/ann/activation_functions/isrlu_function.cpp [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+      y(i) = Fn(x(i));
+  }
+
+  /**
+   * Computes the inverse of the ISRU function
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/activation_functions/isrlu_function.cpp]: Missing stop at the end.

File: src/mlpack/methods/ann/activation_functions/isrlu_function.cpp [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+  static double Inv(const double y, const double alpha = 0.1)
+  {
+    if (y<0)
+      return y / std::sqrt(1 - alpha * y * y);
+    else
+      return y;
+  }
+
+  /**
+   * Computes the inverse of the ISRLU function
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/activation_functions/isrlu_function.cpp]: Missing stop at the end.

File: src/mlpack/methods/ann/activation_functions/isrlu_function.cpp [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+   * Computes the ISRLU function
+   * 
+   * @param x Input data.
+   * @param alpha paramter, default value = 0.1
+   * @return f(x)
+   */
+  static double Fn(const double x, const double alpha = 0.1)
+  {
+    if (x<0)
+      return x / (std::sqrt(1 + alpha*x*x));
```
Comment (prince776) [inline src/mlpack/methods/ann/activation_functions/isrlu_function.cpp]: There should be space between operators and operands in `alpha*x*x`

File: src/mlpack/methods/ann/activation_functions/isrlu_function.cpp [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+  /**
+   * Computes the inverse of the ISRU function
+   * 
+   * @param y
+   * @param alpha
+   * @return f^{-1}(y)
+   */
+  static double Inv(const double y, const double alpha = 0.1)
+  {
+    if (y<0)
```
Comment (prince776) [inline src/mlpack/methods/ann/activation_functions/isrlu_function.cpp]: Space required same as above.

File: src/mlpack/methods/ann/activation_functions/isrlu_function.cpp [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+      return y / std::sqrt(1 - alpha * y * y);
```
Comment (prince776) [inline src/mlpack/methods/ann/activation_functions/isrlu_function.cpp]: Here if `y = 1/√alpha` then it will become y / 0, please add edge case check here.

File: src/mlpack/methods/ann/activation_functions/isrlu_function.cpp [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+   * @param x The resulting inverse of the input data
+   * @param alpha parameter, default value = 0.1
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Inv(const InputVecType& y,
+                  OutputVecType& x,
+                  const double alpha = 0.1)
+  {
+    if (y<0)
+      x = y / arma::sqrt(1 - alpha * arma::pow(y, 2));
```
Comment (prince776) [inline src/mlpack/methods/ann/activation_functions/isrlu_function.cpp]: Same as above, please prevent division by zero.

File: src/mlpack/methods/ann/activation_functions/isrlu_function.cpp [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+
+ public:
+  /**
+   * Computes the ISRLU function
+   * 
+   * @param x Input data.
+   * @param alpha paramter, default value = 0.1
+   * @return f(x)
+   */
+  static double Fn(const double x, const double alpha = 0.1)
```
Comment (zoq) [inline src/mlpack/methods/ann/activation_functions/isrlu_function.cpp]: Can you implement this as a layer, similar to: https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/layer/leaky_relu.hpp, because this function is most likely used in combination with the base layer, and there is no simple way to use a different alpha value.

File: src/mlpack/methods/ann/activation_functions/isrlu_function.cpp [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+  /**
+   * Computes the first derivate of the ISRLU function
+   * 
+   * @param y Input activation
+   * @param alpha parameter, default value = 0.1
+   * @return f'(x) where f(x) = y
+   */
+  static double Deriv(const double y, const double alpha = 0.1)
+  {
+    double x = Inv( y, alpha);
```
Comment (birm) [inline src/mlpack/methods/ann/activation_functions/isrlu_function.cpp]: remove the space after `(`

File: src/mlpack/methods/ann/activation_functions/isrlu_function.cpp [hunk 14]
```diff
@@ -0,0 +1,159 @@
+/**
+ * @file isrlu_function.hpp
```
Comment (birm) [inline src/mlpack/methods/ann/activation_functions/isrlu_function.cpp]: You've mistakenly made this a .cpp file instead of a .hpp file

File: src/mlpack/methods/ann/activation_functions/isrlu_function.cpp [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+/**
+ * The ISRU function, defined by
+ * 
+ * @f{eqnarray*}{
+ * f(x) &=& x/sqrt(1 + alpha*x*x) if x<0
+ *      &=& x                     if x>=0
+ *
+ * f'(x) &=& cube(1/sqrt(1 + alpha*x*x)) if x<0
+ *       &=& 1                           if x>=0
+ * @f
```
Comment (birm) [inline src/mlpack/methods/ann/activation_functions/isrlu_function.cpp]: I think the doxygen end command here is @f} to close the curly brace started on line 23

File: src/mlpack/methods/ann/activation_functions/isrlu_function.cpp (hunk 11 above)
Comment (PranavReddyP16) [reply src/mlpack/methods/ann/activation_functions/isrlu_function.cpp]: How do you think I should handle this? Should I throw an error saying y cannot be 1/sqrt(alpha)?

Comment (kartikdutt18) [reply src/mlpack/methods/ann/activation_functions/isrlu_function.cpp]: One way of handling it would be assigning DBL_MAX. I think I suggested something similar in PR of @prince776. I will take a closer look at the math for the same, Till then you should convert this implementation into layer.

File: src/mlpack/methods/ann/activation_functions/isrlu_function.cpp (hunk 3 above)
Comment (PranavReddyP16) [reply src/mlpack/methods/ann/activation_functions/isrlu_function.cpp]: Hey @kartikdutt18 the spacing seems to be the same in other activation functions as well. Are you referring to indentation maybe?

Comment (kartikdutt18) [reply src/mlpack/methods/ann/activation_functions/isrlu_function.cpp]: Yes. Arma should align with const

File: src/mlpack/methods/ann/activation_functions/isrlu_function.cpp (hunk 11 above)
Comment (PranavReddyP16) [reply src/mlpack/methods/ann/activation_functions/isrlu_function.cpp]: Ok I will do that thanks

File: src/mlpack/methods/ann/activation_functions/isrlu_function.cpp (hunk 3 above)
Comment (PranavReddyP16) [reply src/mlpack/methods/ann/activation_functions/isrlu_function.cpp]: Oh I'm sorry. I misread indentation guidelines

File: src/mlpack/methods/ann/activation_functions/isrlu_function.cpp (hunk 15 above)
Comment (zoq) [reply src/mlpack/methods/ann/activation_functions/isrlu_function.cpp]: It's `@f[`.

File: src/mlpack/methods/ann/activation_functions/isrlu_function.cpp (hunk 11 above)
Comment (PranavReddyP16) [reply src/mlpack/methods/ann/activation_functions/isrlu_function.cpp]: Hey @kartikdutt18 I have converted to a layer. Can you take a look at it. Also any update about the edge case?

File: src/mlpack/methods/ann/layer/isrlu.hpp [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+  /**
+   * Computes the first derivative of the ISRLU function.
+   *
+   * @param x Input data.
+   * @return f'(x)
+   */
+  double Deriv(const double x)
+  {
+    return (x >= 0) ? 1 : std::pow
+    (1 / std::sqrt(1 + alpha * std::pow(x, 2)), 3);
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/layer/isrlu.hpp]: I would prefer it if you used same syntax everywhere either `x * x or pow(x, 2)` whichever you think is better. Also you need to add the edge case here also.

File: src/mlpack/methods/ann/layer/isrlu.hpp [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+    typename InputDataType = arma::mat,
+    typename OutputDataType = arma::mat
+>
+class ISRLU
+{
+ public:
+  /**
+   * Create the ISRLU object using the specified parameters.
+   * The non zero gradient can be adjusted by specifying the parameter
+   * alpha in the range 0 to 1. Default (alpha = 0.03)
```
[the remark this answers is not in the document]
Comment (kartikdutt18) [reply src/mlpack/methods/ann/layer/isrlu.hpp]: Also if alpha can be negative write it as -1 to 1  or magnitude of alpha between 0 and 1.

File: src/mlpack/methods/ann/layer/isrlu_impl.hpp [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+#include "isrlu.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+ISRLU<InputDataType, OutputDataType>::ISRLU(
+    const double alpha) : alpha(alpha)
+{
+  // Nothing to do here.
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/layer/isrlu_impl.hpp]: I think it would more sense to have an assert statement here that ensures alpha is between -1 and 1.

**Comment by kartikdutt18** [comment]
You need to add the layer to CMake list in ann/layer. You can locally test the build using sudo make -jN where N is the number of cores you want to use. To test mlpack use sudo make test -jN 
I have also pointed out cause of build error above. Kindly have a look. Thanks for all your work.

File: src/mlpack/methods/ann/layer/isrlu.hpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (PranavReddyP16) [reply src/mlpack/methods/ann/layer/isrlu.hpp]: Hey @kartikdutt18 This is something I seemed to have missed. Alpha can be negative but this could lead to an error if x < -1/sqrt(alpha) since we would then be trying to find square root of a negative number. How do you think I should define the range of alpha over here?

[the remark this answers is not in the document]
Comment (kartikdutt18) [reply src/mlpack/methods/ann/layer/isrlu.hpp]: I think domain of x would be - 1 / sqrt(alpha) to 1 (or infinity)

[the remark this answers is not in the document]
Comment (kartikdutt18) [reply src/mlpack/methods/ann/layer/isrlu.hpp]: Could you add a link to the documentation if you have any. I would also find something to clear this up. Thanks.

[the remark this answers is not in the document]
Comment (PranavReddyP16) [reply src/mlpack/methods/ann/layer/isrlu.hpp]: Yes that's the domain of x. Should I specify that also somewhere? And what should the range of alpha be specified as?

[the remark this answers is not in the document]
Comment (kartikdutt18) [reply src/mlpack/methods/ann/layer/isrlu.hpp]: Add a fatal error for that. Use `Log:: Fatal <<" Error message goes here."<< std::endl;`

**Comment by PranavReddyP16** [comment]
@kartikdutt18 should I also add tests for this?

**Comment by PranavReddyP16** [comment]
Also I have fixed everything except for the ranges

**Comment by kartikdutt18** [comment]
I would first go about working on ranges and then add tests. Since there are various points of failures (edge cases) could you also add test cases for the same. Thanks.

**Comment by PranavReddyP16** [comment]
Hey @kartikdutt18 the edge cases occur only when the condition alpha <= -1 / x^2. As long this is true for the greatest value of x, it will be true for all values of x. So should I just limit alpha to be greater than zero or should I warn the user that this will happen in case the value of alpha is too small

File: src/mlpack/methods/ann/activation_functions/CMakeLists.txt [hunk 19]
```diff
@@ -11,6 +11,7 @@ set(SOURCES
   mish_function.hpp
   lisht_function.hpp
   gelu_function.hpp
+  isrlu_function.hpp
```
[the remark this answers is not in the document]
Comment (kartikdutt18) [reply src/mlpack/methods/ann/activation_functions/CMakeLists.txt]: This needs to be put in ann/layer/CMakeLists.txt 
This is causing build errors.

**Comment by PranavReddyP16** [comment]
sorry looks like I forgot to include that file when I did git add

**Comment by PranavReddyP16** [comment]
Hey @kartikdutt18 any clue as to why all the windows related tests are failing?

**Comment by kartikdutt18** [comment]
It's unrelated to your PR. mlpack recently switched servers (so some tests including style etc aren't available) Don't worry about them. I  think they will be up and running soon.

**Comment by PranavReddyP16** [comment]
Ok Thanks :)

**Comment by kartikdutt18** [comment]
Could you also add tests for the same. Thanks.

**Comment by PranavReddyP16** [comment]
Hey @kartikdutt18 can you tell me a bit about where to get started while adding tests to this? I've tried going through the codebase but didn't really understand what was happening. If there are any links regarding these I can read those too.

**Comment by kartikdutt18** [comment]
Create a BOOST_AUTO_TEST_CASE(testcaseName) function.
Create desired value matrices for the input at the top of the file. We have a function implemented to check their equality. You can refer any other test case for the same.
Hope this helps.

**Comment by PranavReddyP16** [comment]
@kartikdutt18 I'm getting an error at the following line in the file mlpack/src/mlpack/tests/activation_functions_test.cpp:

https://github.com/mlpack/mlpack/blob/cfaab7877bb7498d18198127d39c3f177fe2d11b/src/mlpack/tests/activation_functions_test.cpp#L25
https://github.com/mlpack/mlpack/blob/cfaab7877bb7498d18198127d39c3f177fe2d11b/src/mlpack/tests/activation_functions_test.cpp#L26

the error is:
mlpack/methods/ann/activation_functions/lisht_function.hpp: No such file or directory
mlpack/methods/ann/activation_functions/gelu_function.hpp: No such file or directory

I have not made any changes to the file at all and the above files are also present in the activation_functions directory. Any reason these errors are being thrown?

**Comment by kartikdutt18** [comment]
If you are using vscode, It does this for libraries not installed.After you updated your branch, it got some new functionalities, however I'm assuming you didn't run the installation command. To fix this use `sudo make install -j1`.(replace 1 with cores you want to use for the same).

**Comment by PranavReddyP16** [comment]
Currently I'm working on vim but I'll give it a try anyway

**Comment by kartikdutt18** [comment]
Hi @PranavReddyP16, could you modify your code in accordance to this conversation. https://github.com//pull/2174/files#r384141238
In essence, remove `fn `and `deriv `from the code implement them directly in Forward and Backward.
Thanks.

**Comment by rcurtin** [comment]
This code touches the ANN code, and there was recently a big refactoring of the ANN code in #2259, so be sure to merge the master branch into your branch here to make sure that nothing will fail if this PR is merged. :+1: (I'm pasting this message into all possibly relevant PRs.)

File: src/mlpack/methods/ann/layer/isrlu.hpp [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+#ifndef MLPACK_METHODS_ANN_LAYER_ISRLU_HPP
+#define MLPACK_METHODS_ANN_LAYER_ISRLU_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * The ISRLU function, also known as Inverse Square Root Linear Unit
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/isrlu.hpp:19]:
```suggestion
 * The ISRLU function, also known as Inverse Square Root Linear Unit.
```

File: src/mlpack/methods/ann/layer/isrlu.hpp [hunk 21]
```diff
@@ -0,0 +1,181 @@
...
+ * It has a negative value allowing it to push the mean unit activation closer
+ * to zero and bring the normal gradient closer to the unit natural gradient,
+ * ensuring a noise robust deactivation state, lessening the over fitting
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/isrlu.hpp:22]:
```suggestion
 * ensuring a noise robust deactivation state, lessening the overfitting
```

File: src/mlpack/methods/ann/layer/isrlu.hpp [hunk 22]
```diff
@@ -0,0 +1,181 @@
...
+ * risk. Experiments have led to the conclusion that ISRLU leads to faster
+ * learning and better generalization on CNN's than ReLU.
+ *
+ * It is defined by
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/isrlu.hpp:26]: Can you cite https://arxiv.org/abs/1710.09967 here as well, here is an example how the format looks like:

https://github.com/mlpack/mlpack/blob/fded9ee333bfbac179e9826bdfe4a3b093ef67f2/src/mlpack/methods/ann/layer/elu.hpp#L52-L62

File: src/mlpack/methods/ann/layer/isrlu.hpp [hunk 23]
```diff
@@ -0,0 +1,181 @@
...
+    typename InputDataType = arma::mat,
+    typename OutputDataType = arma::mat
+>
+class ISRLU
+{
+ public:
+  /**
+   * Create the ISRLU object using the specified parameters.
+   * The non zero gradient can be adjusted by specifying the parameter
+   * alpha in the range 0 to 1. Default value of alpha = 0.03
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/isrlu.hpp:50]:
```suggestion
   * alpha in the range 0 to 1. Default value of alpha = 0.03.
```
